src/psyclone/domain/lfric/kernel/common_arg.py

```diff
@@ -0,0 +1,196 @@
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford, STFC Daresbury Lab
+
+'''Module containing the abstract CommonArg class which captures the
+metadata associated with a generic LFRic argument. Supports the
+creation, modification and Fortran output of such an argument.
+
+'''
+from abc import ABC, abstractmethod
+
+from fparser.common.readfortran import FortranStringReader
+from fparser.two import Fortran2003
+from fparser.two.parser import ParserFactory
+
+
+class CommonArg(ABC):
+    '''Abstract class to capture common LFRic kernel metadata.
+
+    :param Optional[str] datatype: the datatype of this argument.
+    :param Optional[str] access: the way the kernel accesses this \
+        argument.
+
+    '''
+    def __init__(self, datatype=None, access=None):
+        self._form = None
+        if datatype is None:
+            self._datatype = datatype
+        else:
+            self.datatype = datatype
+        if access is None:
+            self._access = access
+        else:
+            self.access = access
+
+    @staticmethod
+    def create_psyir(fortran_string, encoding=Fortran2003.Part_Ref):
+        '''Creates psyir from a Fortran string. In this case the PSyiR will be
+        encoded as an fparser2 tree. The resultant parent node of the
+        tree will be a Part_Ref or a Structure_Constructor object (as
+        it represents a metadata argument).
+
+        :param str fortran_string: a string containing the metadata in \
+           Fortran.
+        :param encoding: the parent class with which we will encode the \
+            Fortran string. This is an optional argument that defaults to \
+            Fortran2003.Part_Ref.
+        :type encoding: Optional[type]
+
+        :returns: an fparser2 tree containing the PSyIR for a metadata \
+            argument.
+        :rtype: :py:class:`fparser.two.Fortran2003.Part_Ref` | \
+            :py:class:`fparser.two.Fortran2003.Structure_Constructor`
+
+        :raises ValueError: if the Fortran string is not in the \
+            expected form.
+
+        '''
+        _ = ParserFactory().create(std="f2003")
+        reader = FortranStringReader(fortran_string)
+        psyir = encoding(reader)
+        if not psyir:
+            raise ValueError(
+                f"Expected kernel metadata to be a Fortran "
+                f"{encoding.__name__}, with the form 'arg_type(...)' "
+                f"but found '{fortran_string}'.")
+        return psyir
+
+    @staticmethod
+    def check_psyir(psyir, nargs=4, encoding=Fortran2003.Part_Ref):
+        '''Checks that the PSyIR is valid. The PSyIR may have a
+        different number of arguments depending on its type and it may
+        be in the form of a Fortran2003 Part_Ref or Fortran2003
+        Structure_Constructor.
+
+        :param psyir: fparser2 tree containing the PSyIR for a \
```

Comment:
I don't understand how an fparser2 parse tree can contain PSyIR?

Reply:
representing

Reply:
This argument has been renamed.